src/cata_tiles.cpp

```diff
@@ -680,43 +716,72 @@ tile_type &cata_tiles::load_tile(JsonObject &entry, const std::string &id, int o
 {
     tile_type &curr_subtile = tile_ids[id];
 
-    if ( entry.has_array("fg") ) {
-        JsonArray fg_array = entry.get_array("fg");
-        while (fg_array.has_more()) {
-            const int fg = fg_array.next_int();
-            if ( fg >= 0 ) {
-                curr_subtile.fg.push_back(fg);
+    // loop through fg and bg
+    for (auto g : {std::make_pair("fg",&curr_subtile.fg), std::make_pair("bg",&curr_subtile.bg)}) {
+        // one sprite_variation for bookkeeping purposes
+        g.second->push_back(sprite_variation());
+        g.second->at(0).weight = -1;
+        // json array indicates rotations or variations
+        if ( entry.has_array(g.first) ) {
+            JsonArray g_array = entry.get_array(g.first);
+            // int elements of array indicates rotations
+            // create one variation, populate sprite_ids with list of ints
+            if ( g_array.test_int() ) {
+                g.second->push_back(sprite_variation());
+                while (g_array.has_more()) {
+                    const int s = g_array.next_int();
+                    if ( s >= 0 ) {
+                        g.second->at(0).sprite_ids.push_back(s);
+                    }
+                }
             }
-        }
-    } else if (entry.has_int("fg") && entry.get_int("fg") >= 0) {
-        curr_subtile.fg.push_back(entry.get_int("fg"));
-    }
-
-    if ( entry.has_array("bg") ) {
-        JsonArray bg_array = entry.get_array("bg");
-        while (bg_array.has_more()) {
-            const int bg = bg_array.next_int();
-            if ( bg >= 0 ) {
-                curr_subtile.bg.push_back(bg);
+            // object elements of array indicates variations
+            // create one variation per object
+            else if ( g_array.test_object() ) {
+                while (g_array.has_more()) {
+                    sprite_variation v = sprite_variation();
+                    JsonObject vo = g_array.next_object();
+                    v.weight = vo.get_int("weight");
+                    // negative weight is invalid
+                    if ( v.weight < 0 ) {
+                        vo.throw_error("Invalid weight for sprite variation (<0)", g.first);
+                        continue;
```

Comment:
> the "user" here is the player, not the tileset author. I do not want players to get kicked out of the game because they tried to load an invalid tileset.

I would have assumed that a tileset author tries their own tileset at least once. That's when the error message comes up and forces them to fix it. Anthe incorrect tileset won't even get published in the first place.

I can, however, understand your reasoning.